### ISSUE TITLE
Observe + prune throws an error

### ISSUE THREAD

**Issue by yuriy0** [hakaru-dev/hakaru#105]
Consider the following simple program:

```
x <~ dirac 1
y <~ plate i of 3: observe(uniform(0,1), 1)
return ()
```

After evaluating `observe` with `expandTransformations`, the result is

```
syn (MBind :$ 
    ((syn (Dirac :$ ((syn (Literal_ (LNat 1))) :* End))) :* 
    ((bind (Variable "x" (Nat 0) SNat) -- 1
           (syn (MBind :$ 
                ((syn (Plate :$ ((syn (Literal_ (LNat 3))) :* 
                      ((bind (Variable "i" (Nat 2) SNat) 
                             (syn (Let_ :$ ((syn ((CoerceTo_ ...) :$ 
                                  ((syn (Literal_ (LNat 1))) :* End))) :* 
                                  ((bind (Variable "" (Nat 0) SReal) ...) :* End))))) :* End)))) :* -- 2
                ((bind (Variable "y" (Nat 1) (SArray SReal)) 
                       (syn (Dirac :$ ...))) :* End))))) :* End)))
```
(Irrelevant parts omitted). Note lines marked 1 and 2 - the variable `0` is bound on line 1 and bound again in an inner scope on line 2. I believe this is technically a valid program, because the `...` expression on line 2 is free of the outer variable (so the variable is shadowed, if that constitutes a valid Hakaru program). 

However, a call to `prune` ([here](https://github.com/hakaru-dev/hakaru/blob/bbf5b116e8ebef62447cd9a83e95a62109d712b1/haskell/Language/Hakaru/Syntax/AST/Transforms.hs#L210)) will then fail. As it traverses the syntax tree, it collects variables in a set, and at some point compares two variables for equality. These two variables have the same ID, but different types, which causes a type error to be raised.

I don't know which part of this program (`observe` or `prune`) is really responsible for this error. Either `observe` should be responsible for producing a program which binds 'new' variables which are fresh in the context in which it appears; or `prune` should be using a different interpretation of variable equality.

Note the output of converting the given program to internal syntax is the same on `master` and `transforms_as_syntax`, but `prune` (and presumably other functions which call `prune`, e.g. `optimizations`) is not called on such programs on `master`.

**Comment by zaxtax** [comment]
This looks like a bug in `observe`

**Comment by ccshan** [comment]
Yeah, it seems that `runDis` needs to receive from its calling context the maximum variable ID used in the rest of the program (or at least bound in the current scope), rather than just using `maxNextFree es`. Doesn't `sendToMaple` need the same context information?

**Comment by JacquesCarette** [comment]
Is there are natural owner for this issue?

**Comment by yuriy0** [comment]
I am working on this; the prospective solution being for `runDis` and other transformations to "receive from its calling context the maximum variable ID".

**Comment by yuriy0** [comment]
I believe this is done as of 4afca6301ac82788e6fef66046b7d8cbd3f912ce.

When I (that is, the CI build) have made sure there are no compatibility issues on GHC 7.8 (as I've still not figured out a way to get that version to work on my machine...) I'll merge this work into master.

**Comment by yuriy0** [comment]
The relevant work has been merged into master (0a960e679a40054416741829cfeb6b5b0974ddb1, I believe).
